File: pycom/timing/scripts/cpu/time_sweep_fdfd_cpu.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov  5 16:14:54 2019

@author: aweiss
"""

from pycom.timing.OperationTimer import fancy_timeit_matrix_sweep
from pycom.timing.FDFD.FDFD_2D import FDFD_2D
import numpy as np
import scipy.linalg
import os

#this allows setting of a global debug
try:
    from __main__ import DEBUG
    if DEBUG is None:
        raise Exception("DEBUG is None")
except:
    DEBUG = True #debug mode for testing
    print("DEBUG Not defined. Setting to False")
if DEBUG:
    print("Debugging Mode Enabled")
    
# what directory to store the data in
try:
    from __main__ import output_directory
except:
    output_directory = './data'

#%% Some initialization
if DEBUG:
    dim_list = np.floor(np.linspace(1,50,6)).astype(np.uint32);
    num_reps = 10
else:
    dim_list = np.floor(np.linspace(1,500,51)).astype(np.uint32);
    num_reps = 100
    
out_file = os.path.join(output_directory,'stats_py_np_fdfd_single.mat')

num_reps = 100
#num_reps = 10

stats_np_single = {}
stats_np_double = {}

#argument generation
def fdfd_arg_gen_funct(dim,num_args): #generate num_cells_x,num_cells_y as dim
    return [dim,dim,np.cdouble]  
def fdfd_arg_gen_funct_single(dim,num_args): #generate num_cells_x,num_cells_y as dim
    return [dim,dim,np.csingle]  

#now run
[solve_stats_double,_] = fancy_timeit_matrix_sweep(
        [FDFD_2D],['fdfd'],[3],dim_list,num_reps
        ,arg_gen_funct=fdfd_arg_gen_funct);
[solve_stats_single,_] = fancy_timeit_matrix_sweep(
        [FDFD_2D],['fdfd'],[3],dim_list,num_reps
        ,arg_gen_funct=fdfd_arg_gen_funct_single);
		
#stats_np_double.update(solve_stats_double)
stats_np_single.update(solve_stats_single)
    
#%% now save out (single is not used)
import scipy.io as sio
sio.savemat(out_file,{'single':stats_np_single,'double':stats_np_double})
    

File: FancyTimer/examples/scripts/cpu/time_sweep_fdfd_cpu.m
path_to_pycom = 'A:\git\';
addpath(fullfile(path_to_pycom,'pycom\python_matlab_comparison\FDFD\'));
%% some setup
if ~exist('pycom_path','var') % check for pycom path
    file_dir = fileparts(mfilename('fullpath')); %get the directory of this file
    pycom_path = fullfile(file_dir,'../../'); %assume the base is 2 levels up from this
end

if ~exist('DEBUG','var') % Check for debug flag
    DEBUG = false;
    fprintf("DEBUG Not defined. Setting to False");
end

if ~exist('output_directory','var') % check for provided output directory
    output_directory = './data';
end

addpath(fullfile(pycom_path,'base')); % add OperationTimer
addpath(fullfile(pycom_path,'python_matlab_comparison\FDFD\'));

%% Some initialization
if DEBUG
    dim_list = floor(linspace(1,50,6));
    num_reps = 5;
else
    dim_list = floor(linspace(1,500,51));
    num_reps = 100;
end
out_file = fullfile(output_directory,'stats_mat_fdfd.mat');

%% Test the fancy_timeit_matrix_sweep capability
functs      = {@FDFD_2D};
funct_names = {'fdfd'  };
num_args    = [2      ];

% now run 
stats_mat_double = OperationTimer.fancy_timeit_matrix_sweep(...
	functs,funct_names,num_args,dim_list,num_reps,'arg_gen_funct',@fdfd_arg_gen_funct);
	
%% now save out
m_single = struct(); % matlab cant do sparse single
m_double = stats_mat_double;
save(out_file,'m_single','m_double');
    
%% functions for fft arg generation                                     
function args = fdfd_arg_gen_funct(dim,num_args)
     %@brief generate default arguments
    args = {dim,dim};
end

function new_struct = update_struct(struct_to_update,struct_to_add)
    %@brief Add all of the fields of struct_to_add to struct_to_update
    %@note If overlapping fields exist, struct_to_update will be overwritten
    %@param[in] struct_to_update - what struct are we updating
    %@param[in] struct_to_add - struct with fields to add to struct_to_update
    %@return New structure with fields from both input structs
    sta_fldnames = fieldnames(struct_to_add);
    new_struct = struct_to_update; %copy struct to update
    for i=1:length(sta_fldnames) %add all of our fieldnames
        fldname = sta_fldnames{i};
        new_struct.(fldname) = struct_to_add.(fldname); %add to new struct
    end
end
    
    
    
    
	
	
	